### PR TITLE
profile refactor on BE

### PR DESCRIPTION
**Pull request by zhiqiang-hhhh**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/runtime/fragment_mgr.cpp [hunk 1]
```diff
@@ -1609,4 +1620,35 @@ void FragmentMgr::get_runtime_query_info(std::vector<WorkloadQueryInfo>* query_i
...
+//     if (iter == _pipeline_map.end()) {
+//         return nullptr;
+//     }
+
+//     return iter->second->collect_profile_p();
+// }
+
+std::tuple<TUniqueId, TNetworkAddress,
+           std::unordered_map<int, std::vector<std::shared_ptr<TRuntimeProfileTree>>>>
+FragmentMgr::collect_profile_x(const TUniqueId& query_id) {
```
Comment (yiguolei) [inline be/src/runtime/fragment_mgr.cpp]: add std::unordered_map<int, std::vector<std::shared_ptr<TRuntimeProfileTree as output parameter.
Return status.

File: be/src/runtime/query_context.cpp [hunk 2]
```diff
@@ -293,4 +310,113 @@ Status QueryContext::set_workload_group(WorkloadGroupPtr& tg) {
...
+#ifndef NDEBUG
+    for (const auto& p : pipeline_profiles) {
+        DCHECK(p != nullptr) << fmt::format("Add pipeline profile failed, query {}, fragment {}",
+                                            print_id(this->_query_id), f_id);
+    }
+#endif
+    _profile_map_x->insert(std::make_pair(f_id, pipeline_profiles));
+}
+
+void QueryContext::add_instance_profile(const TUniqueId& i_id,
```
Comment (yiguolei) [inline be/src/runtime/query_context.cpp]: instance_id

File: be/src/runtime/query_context.cpp [hunk 3]
```diff
@@ -293,4 +310,113 @@ Status QueryContext::set_workload_group(WorkloadGroupPtr& tg) {
...
+QueryContext::collect_profile_x() {
+    auto res = std::make_tuple(
+            this->_query_id, this->coord_addr,
+            std::unordered_map<int, std::vector<std::shared_ptr<TRuntimeProfileTree>>>());
+    auto& profile_map = std::get<2>(res);
+
+    // std::map<int, std::weak_ptr<pipeline::PipelineFragmentContext>>
+    for (auto [f_id, f_profile_wptr] : _fragment_id_to_pipeline_ctx) {
+        if (auto f_profile = f_profile_wptr.lock()) {
+            profile_map[f_id] = f_profile->collect_profile_x();
```
Comment (yiguolei) [inline be/src/runtime/query_context.cpp]: rename to fragment_id

File: be/src/runtime/query_context.cpp [hunk 4]
```diff
@@ -293,4 +310,113 @@ Status QueryContext::set_workload_group(WorkloadGroupPtr& tg) {
...
+    std::lock_guard<std::mutex> lg(_profile_mutex);
+    LOG_INFO("Query {}, register query profile, instance profile count {}", print_id(_query_id),
+             _profile_map->size());
+
+    for (auto& [i_id, i_profile] : *_profile_map) {
+        ExecEnv::GetInstance()->runtime_query_statistics_mgr()->register_instance_profile(
+                _query_id, this->coord_addr, i_id, i_profile);
+    }
+
+    _profile_map->clear();
```
Comment (yiguolei) [inline be/src/runtime/query_context.cpp]: not call clear

File: be/src/pipeline/pipeline_fragment_context.cpp [hunk 5]
```diff
@@ -993,4 +995,50 @@ std::string PipelineFragmentContext::debug_string() {
...
+        const {
+    std::vector<std::shared_ptr<TRuntimeProfileTree>> res;
+    DCHECK(_query_ctx->enable_pipeline_x_exec() == true)
+            << fmt::format("Query {} calling a pipeline X function, but its pipeline X is disabled",
+                           print_id(this->_query_id));
+    std::stringstream ss;
+
+    for (auto& pipeline_profile : _runtime_state->pipeline_id_to_profile()) {
+        auto profile_ptr = std::make_shared<TRuntimeProfileTree>();
+        pipeline_profile->to_thrift(&(*profile_ptr));
```
Comment (yiguolei) [inline be/src/pipeline/pipeline_fragment_context.cpp]: &(*profile_ptr)     ---》   profile_ptr.get

File: be/src/pipeline/pipeline_fragment_context.cpp [hunk 6]
```diff
@@ -993,4 +995,50 @@ std::string PipelineFragmentContext::debug_string() {
...
+            print_id(this->_query_id));
+
+    // std::stringstream ss;
+    DCHECK(_runtime_state->runtime_profile() != nullptr);
+    _runtime_state->runtime_profile()->compute_time_in_profile();
+    // _runtime_state->runtime_profile()->pretty_print(&ss);
+    _runtime_state->runtime_profile()->to_thrift(&(*res));
+
+    return res;
+}
```
Comment (yiguolei) [inline be/src/pipeline/pipeline_fragment_context.cpp]: Should also add method in plan fragment executor

File: be/src/runtime/fragment_mgr.cpp [hunk 7]
```diff
@@ -610,6 +611,16 @@ void FragmentMgr::remove_pipeline_context(
             _pipeline_map.erase(ins_id);
             g_pipeline_fragment_instances_count << -1;
         }
+
+        auto q_context = _query_ctx_map.find(query_id)->second;
```
Comment (yiguolei) [inline be/src/runtime/fragment_mgr.cpp]: should also add plan_fragment_executor's profile to query ctx

File: be/src/pipeline/pipeline_fragment_context.cpp (hunk 6 above)
Comment (RoanHeNaN) [reply be/src/pipeline/pipeline_fragment_context.cpp]: > Should also add method in plan fragment executor

done
https://github.com/yiguolei/incubator-doris/pull/4/files#diff-03276016fa04b08d068c8eab318f3529bf0d153c0e23e7e006d4f790f2506efe

File: be/src/runtime/runtime_query_statistics_mgr.cpp [hunk 8]
```diff
@@ -17,14 +17,342 @@
...
+#include "runtime/query_context.h"
+#include "service/backend_options.h"
 #include "util/debug_util.h"
 #include "util/time.h"
+#include "util/uid_util.h"
 #include "vec/core/block.h"
 
 namespace doris {
 
+static Status _doReportExecStatsRpc(const TNetworkAddress& coor_addr,
```
Comment (yiguolei) [inline be/src/runtime/runtime_query_statistics_mgr.cpp:48]: change to lowercase

File: be/src/runtime/runtime_query_statistics_mgr.cpp [hunk 9]
```diff
@@ -17,14 +17,342 @@
...
+                req.__set_loadChannelProfile(*load_channel_profile_copy[instance_id]);
+            }
+
+            TReportExecStatusResult res;
+            auto rpc_status = _doReportExecStatsRpc(coor_addr, req, res);
+
+            if (res.status.status_code != TStatusCode::OK) {
+                std::stringstream ss;
+                res.status.printTo(ss);
+                LOG_WARNING("Query {} send profile to {} failed, msg: {}", print_id(query_id),
```
Comment (yiguolei) [inline be/src/runtime/runtime_query_statistics_mgr.cpp:293]: Also print instance id

File: be/src/runtime/runtime_query_statistics_mgr.cpp [hunk 10]
```diff
@@ -17,14 +17,342 @@
...
+                            PrintThriftNetworkAddress(coor_addr), ss.str());
+            } else {
+                LOG_INFO("Send {} profile finished", print_id(query_id));
+            }
```
Comment (yiguolei) [inline be/src/runtime/runtime_query_statistics_mgr.cpp:297]: print query id and instance id

File: be/src/runtime/runtime_query_statistics_mgr.cpp [hunk 11]
```diff
@@ -17,14 +17,342 @@
...
+            TReportExecStatusParams req = create_report_exec_status_params(
+                    query_id, fragment_id, fragment_profile, load_channel_profile);
+            TReportExecStatusResult res;
+
+            auto rpc_status = _doReportExecStatsRpc(coor_addr, req, res);
+
+            if (res.status.status_code != TStatusCode::OK ||
+                res.status.status_code != TStatusCode::OK) {
+                LOG_WARNING("Query {} send profile to {} failed", print_id(req.query_id),
+                            PrintThriftNetworkAddress(coor_addr));
```
Comment (yiguolei) [inline be/src/runtime/runtime_query_statistics_mgr.cpp:348]: print fragment id

**Pull request closed without merging**